Support PARSENAME() T-SQL function

pratikzode opened the pull request:
The PARSENAME() function in T-SQL is used to parse a string representing a four-part SQL Server object name, such as "database.schema.object.column".

SELECT PARSENAME('tempdb.dbo.Employee', 1) 
2> go                                                                                                                        

Employee   

1> SELECT PARSENAME('tempdb.dbo.Employee', 2) 
2> go                                                                                                                           

dbo          

1> SELECT PARSENAME('tempdb.dbo.Employee', 3) 
2> go                                                                                                                          

tempdb   


Signed-off-by: Pratik Zode [pzode@amazon.com](mailto:pzode@amazon.com)

pratikzode committed:
Support PARSENAME() T-SQL function

The PARSENAME() function in T-SQL is used to parse a string representing a four-part SQL Server object name, such as "database.schema.object.column".

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified Expected file

Removed “Could not find tests for function sys.parsename” ,””Could not find upgrade tests for function sys.parsenamefrom expected_create.out

Removed “Function sys.parsename(character varying,integer)” from expected_dependency.out

Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,135 @@ numeric_radians(PG_FUNCTION_ARGS)
 }
 
+Datum
+parsename(PG_FUNCTION_ARGS) {
```
Please go through the whole change and correct all the indents and code format

xhfanhe commented on test/JDBC/input/BABEL-808-vu-verify.sql:
```diff
@@ -0,0 +1,98 @@
+GO
+
+SELECT PARSENAME('.dbo.Employee',3)
+GO
```
Please add test cases for strings with 4 parts and >4 parts

xhfanhe commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.1.0--3.2.0.sql:
```diff
@@ -824,6 +824,11 @@ CREATE AGGREGATE sys.VARP(float8) (
 CREATE OR REPLACE FUNCTION sys.rowcount_big()
 RETURNS BIGINT AS 'babelfishpg_tsql' LANGUAGE C STABLE;
 
+CREATE OR REPLACE FUNCTION sys.parsename(object_name character varying, object_piece int)
```
Need to do the function renaming stuff like user_id - rename to deprecated and later drop it

xhfanhe commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -336,6 +336,7 @@ BABEL-1249
 BABEL-1291
 BABEL-1994-CHAR
 BABEL-1994-VARCHAR
+BABEL-808
```
Please add the test to all upgrade schedules, starting from 13_4

xhfanhe commented on test/JDBC/input/BABEL-808-vu-prepare.sql:
```diff
@@ -0,0 +1,51 @@
+CREATE TABLE Employee (
```
1. Please put the test scripts to input/functions and name it with function name, like we did for user_id
2. Please change all the table/view/procedure names to include the test title

xhfanhe commented on contrib/babelfishpg_tsql/sql/sys_functions.sql:
```diff
@@ -551,21 +551,10 @@ RETURNS INTEGER AS
-		ELSE NULL
-		END $$ immutable LANGUAGE 'sql';
+CREATE OR REPLACE FUNCTION sys.parsename(object_name character varying, object_piece int)
+RETURNS character varying
```
The function return type should be sys.SYSNAME

pratikzode committed:
Modified sql file

Have modified the return type to sys.SYSNAME.
Added test to input/functions
Added test to all upgrade schedules
Did the alignment for function.c file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Alignment in c file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Removed sys-parsename from upgrade schedules

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Removed Alter function from 3.1-3.2

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added Alter function to sql file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Renaming functions

Added test in sys-parsename-dep and added to all the schedule files

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified ALTER FUNCTION sys.parsename(VARCHAR,INT)

Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented on test/JDBC/input/functions/sys-parsename-vu-verify.sql:
```diff
@@ -0,0 +1,107 @@
+SELECT PARSENAME('.dbo.Employee.table',3)
+GO
+
+SELECT PARSENAME('.dbo.Employee.table',4)
```
Please add a test where the string has four parts, like 
```
SELECT PARSENAME('aaa.bbb.ccc.ddd',4)
```
and a string that has >4 parts, like
```
SELECT PARSENAME('aaa.bbb.ccc.ddd.eee',3)
```
And please make sure the output is the same as SQL server

pratikzode committed:
Added test Cases

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified the C file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Resolved Conflicts

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Merge Conflicts resolved

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified C file added new test cases

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified parsername-dep test case

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified C file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified C file

Have modified c file to handle double quotes

Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/expected/sys-parsename-vu-verify.out:
```diff
@@ -0,0 +1,358 @@
+~~END~~
+
+
+SELECT PARSENAME('tempdb.dbo.Employee.table.mytable',3)
```
this result is wrong

pratikzode replied:
I have given 5 'tempdb.dbo.Employee.table.mytable' is it should be NULL
SQL output:
1>SELECT PARSENAME('tempdb.dbo.Employee.table.mytable',3)
2>GO                                                                                                                      

NULL

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,109 @@ numeric_radians(PG_FUNCTION_ARGS)
+    int inside_brackets = 0;
+    int unmatched_bracket = 0;
+    if (object_piece < 1 || object_piece > 4)
+    {
```
As I mentioned multiple times, please use tab instead spaces.

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,109 @@ numeric_radians(PG_FUNCTION_ARGS)
+    int object_piece = PG_GETARG_INT32(1);
+    char *object_name_str = text_to_cstring(object_name);
+    int len = strlen(object_name_str);
+    char object_parts[5][129] = {0};
```
do we really need this buffer? Can we optimize it out? 
And the TSQL doc mentioned 256B, why it is 129 here? 
Could you please add some test with long unicode ?

macwoody replied:
And is it the right way to initialize a 2d array?

pratikzode replied:
The buffer is used to store the different parts of the object name, so it is necessary

pratikzode replied:
Each part of the 'object_name' string is type sysname which is equivalent to nvarchar(128) or 256 bytes.

forestkeeper replied:
Think it's not a good idea to have a fixed 2-d array for this

forestkeeper replied:
The point is , you don't need to keep all the part from 1 - 4, just get the part function return needs, and it's enough

forestkeeper replied:
And also , you don't need to copy to a target array, just keep a start/end of current part, that'll be enough

pratikzode committed:
Modified the C file and added test cases

Previously the parsername was not giving the as compared to SQL server
but after the modification its giving exact output as desired and have
added new test as well.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

forestkeeper commented on test/JDBC/expected/sys-parsename-dep-vu-prepare.out:
```diff
@@ -0,0 +1,53 @@
+-- Insert some sample data
+INSERT INTO parsename_Employee(EmployeeID, FirstName, LastName, HireDate, Salary)
+VALUES (1, 'John', 'Doe', '2020-01-01', 50000),
+ (2, 'Jane', 'Smith', '2020-02-01', 60000),
```
Should add a test case like : 

SELECT PARSENAME('AdventureWorksPDW2012.dbo.DimCustomer', 1) AS 'Object Name';  
SELECT PARSENAME('AdventureWorksPDW2012.dbo.DimCustomer', 2) AS 'Schema Name';  
SELECT PARSENAME('AdventureWorksPDW2012.dbo.DimCustomer', 3) AS 'Database Name';  
SELECT PARSENAME('AdventureWorksPDW2012.dbo.DimCustomer', 4) AS 'Server Name';  
GO  

and SELECT PARSENAME('AdventureWorksPDW2012.dbo.DimCustomer', 0) AS 'Server Name';   -> ilegal , but need to see what's going on for illegal ones

pratikzode committed:
Added new test cases

Have added more test cases which was documented to get more clarity.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Added test cases to sys-parsename-dep

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified C file

Have modified the code as finite state machine

Signed-off-by: pratikzode <pratik.zode@gmail.com>

xhfanhe commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,172 @@ numeric_radians(PG_FUNCTION_ARGS)
+	int object_piece = PG_GETARG_INT32(1);
+	char *object_name_str = text_to_cstring(object_name);
+	int len = strlen(object_name_str);
+	char object_parts[129 * 5] = {0};
```
As Di and Zhibai mentioned in previous comments, we don't need this buffer. 
The target here is to separate out four parts from the input, so at most we only need to manage four pointers pointing to the start of each part, and record the length of each part. 
In fact, since we know which part needs to be returned, one pointer should be enough.

macwoody replied:
i feel that we may at most need 256 B for one part. 
Please be noticed that according to the doc, one part is up to 256B 
I am not sure why it is 129 x 5 here.

pratikzode replied:
array called "object_parts" that is 129 * 5 in size, meaning it can hold up to 5 elements, each of which is a null-terminated string of maximum length 128 characters. This array is used to store the different parts of an object name that is passed in as an argument to the function

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,172 @@ numeric_radians(PG_FUNCTION_ARGS)
+					state = 0;
+					if (i + 1 < len && object_name_str[i + 1] != '.')
+					{
+						improperly_enclosed = 1;
```
why not error out immediately?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,172 @@ numeric_radians(PG_FUNCTION_ARGS)
+			{
+				state = 2;
+			}
+			else if(c == ']')
```
should double quote be also considered in this state?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,172 @@ numeric_radians(PG_FUNCTION_ARGS)
+			}
+			else
+			{
+				if (part_len < 128)
```
why 128?

pratikzode replied:
If the length of the current part is already 128 characters or more, the function immediately returns a "NULL" value.
This check is done to ensure that the maximum length of each part of the object name does not exceed 128 characters.

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,172 @@ numeric_radians(PG_FUNCTION_ARGS)
+			{
+				if (part_len < 128)
+				{
+					snprintf(&object_parts[current_part * 129] + part_len, 2, "%c", c);
```
what is the use of this snprintf?

pratikzode replied:
The snprint() function is used to format the given strings into a series of characters or values in the buffer area.

macwoody commented:
I got the feeling that since index counted from the right most part, you may consider scanning the input from left to right

pratikzode committed:
Modified the C file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on test/JDBC/upgrade/13_8/schedule:
```diff
@@ -303,6 +303,7 @@ Test-sp_helpsrvrolemember
 Test-sp_helpuser
 TestSQLVariant
 TestTableType
+sys-parsename-dep
```
seems to me the profile is order alphabetically

pratikzode committed:
Modified the function.c file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified upgrade file

Modified upgrade/schedule file in alphabetically order.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified upgrade schedule

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified function.c file

have modified UChar object_parts[129 * 5] = {0}; to UChar object_parts[256] = {0};

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified test case

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified the C file and test cases

Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+	int part_lengths[256] = {0};
+	
+	// Calculate max_parts based on the number of dots in the input string
+	for (char *dot = strchr(object_name_str, '.'); dot != NULL; dot = strchr(dot + 1, '.'))
```
This is wrong, dot is not always delimiter of parts.
It could be part of the name as well. This error should be detected by your test if you add all tests for special characters
1> select parsename('[abc.abc]', 1);
2> go

--------------------------------------------------------------------------------------------------------------------------------
abc.abc

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+		max_parts++;
+	}
+    // Convert the input string to Unicode using ICU
+	u_strFromUTF8(unicode_object_name_str, sizeof(unicode_object_name_str) / sizeof(UChar), NULL, object_name_str, -1, &errorCode);
```
if you would like to work of UTF, you may need to use utility functions already exist in postgres to maintain the compatibility.

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+	// Change the requested object_piece to reverse the order
+	object_piece = max_parts - object_piece + 1;
+	// Determine the maximum allowed length based on whether the input is Unicode or not
+	max_length = (len != strlen(object_name_str)) ? 255 : 128;
```
> Each part of the 'object_name' string is type sysname which is equivalent to nvarchar(128) or 256 bytes. If any part of the string exceeds 256 bytes

Notice: Each PART, not the entire input

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+		{
+			if (c == '"')
+			{
+				state = 1;
```
I think you need to distinguish 2 cases in your state machine 
abc."abcd"  and 
abc.ab"cd" 
The second one is a syntax error

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+			{
+				state = 1;
+			}
+			else if (c == '[')
```
same as above

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+					state = 0;
+					if (i + 1 < len && unicode_object_name_str[i + 1] != '.')
+					{
+						improperly_enclosed = 1;
```
why not terminate from here?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+					state = 0;
+					if (i + 1 < len && unicode_object_name_str[i + 1] != '.')
+					{
+						improperly_enclosed = 1;
```
same as above

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+	int improperly_enclosed = 0;
+	int max_parts = 1;
+	int max_length;
+	int part_lengths[256] = {0};
```
why int array 256?

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+		PG_RETURN_NULL();
+	}
+	// Reset state and current_part for the parsing loop
+	state = 0;
```
please try consider do it in one pass

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2091,237 @@ numeric_radians(PG_FUNCTION_ARGS)
+		// Convert the result from Unicode to UTF-8
+		char utf8_result[256] = {0};
+		text *result;
+		u_strToUTF8(utf8_result, sizeof(utf8_result), NULL, requested_object_part, -1, &errorCode);
```
why do you convert the string again? what would you like to achieve?

pratikzode committed:
Modified C file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified functions.c file

Have taken only one char object_parts[256] = {0}; and used only one
pointer.

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Alignment in C file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,191 @@ numeric_radians(PG_FUNCTION_ARGS)
+			{
+				if(part_len < 128)
+				{
+					*part_current++ = c;
```
I could not understand what this line is doing, could you please explain with an example?

pratikzode replied:
The expression *part_current++ = c; is adding the character c to the current part of the object_parts array and then advancing the part_current pointer to the next character position in the same array

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,191 @@ numeric_radians(PG_FUNCTION_ARGS)
+	}
+	if (improper_close_bracket)
+	{
+		pfree(object_name_str);
```
how could the execution flow reach here?

pratikzode replied:
yes removed it.
It was from earlier code

macwoody commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,191 @@ numeric_radians(PG_FUNCTION_ARGS)
+		part_start = object_parts;
+		for (int i = 0; i < reverse_index; i++)
+		{
+			part_start += strlen(part_start) + 1;
```
could you please help explain how this work ?

pratikzode replied:
The expression part_start += strlen(part_start) + 1; advances the part_start pointer to the next part in the object_parts array.

macwoody commented on test/JDBC/upgrade/latest/schedule:
```diff
@@ -313,6 +313,7 @@ sys-objects
 sys-original_login
 sys-partitions
 sys-partitions-dep
+sys-parsename-dep
```
you have two sets of test sys-parsename-dev and sys-parsename 
have you verified if both will be running ?

pratikzode replied:
YES

pratikzode committed:
Modified Function.c file

Modified the file to hande 128 varchar or 256 byte

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified the upgrade/latest/schedule

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified C file

Added state=0 as well

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Removed the previous commented code

Signed-off-by: pratikzode <pratik.zode@gmail.com>

pratikzode committed:
Modified upgrade test

Have deleted sys-parsename-dep from the test and have added
sys-parsename to all the upgrade schedule files

Signed-off-by: pratikzode <pratik.zode@gmail.com>

lejaokri commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,178 @@ numeric_radians(PG_FUNCTION_ARGS)
+	char *part_start = object_parts;
+	char *part_current = part_start;
+	int current_part = 0;
+	int state = 0; // 0: outside quotes and brackets, 1: inside quotes, 2: inside brackets
```
In the future, suggest using an enum :)

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,178 @@ numeric_radians(PG_FUNCTION_ARGS)
+					state = 0;
+					if (i + 1 < len && object_name_str[i + 1] != '.')
+					{
+						PG_RETURN_NULL();
```
didn't pfree object_name_str in here

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,178 @@ numeric_radians(PG_FUNCTION_ARGS)
+	}
+	else
+	{
+	pfree(object_name_str);
```
indent not correct in here

pratikzode committed:
Modified the upgrade 14_8 and C file

Have removed orderby-before-15_3 from 14_8 schedule file and added pfree(object_name_str); in c file

Signed-off-by: pratikzode <pratik.zode@gmail.com>

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -2088,6 +2089,179 @@ numeric_radians(PG_FUNCTION_ARGS)
+			}
+			else
+			{
+				if (part_len < 128 && (part_current - object_parts) * sizeof(char) < 256)
```
If part_current exceed 256, it'll reach out of object_name_str
